File: lib/services/meeting/utils/meeting_helpers.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'meeting_constants.dart';

/// Helper functions for the meeting service
class MeetingHelpers {
  /// Generate a human-readable meeting ID
  /// Format: {mentorDocId}__{menteeDocId}__{timestamp}
  static String generateMeetingId(String mentorDocId, String menteeDocId, DateTime startTime) {
    final timestamp = startTime.millisecondsSinceEpoch ~/ 1000;
    return '$mentorDocId${MeetingConstants.meetingIdSeparator}$menteeDocId${MeetingConstants.meetingIdSeparator}$timestamp';
  }
  
  /// Parse meeting ID to extract mentor and mentee document IDs
  static Map<String, String>? parseMeetingId(String meetingId) {
    final parts = meetingId.split(MeetingConstants.meetingIdSeparator);
    if (parts.length >= 3) {
      return {
        'mentorDocId': parts[0],
        'menteeDocId': parts[1],
        'timestamp': parts[2],
      };
    }
    return null;
  }
  
  /// Generate availability slot ID
  /// Format: {mentorDocId}__{timestamp}
  static String generateAvailabilityId(String mentorDocId, DateTime slotTime) {
    final timestamp = slotTime.millisecondsSinceEpoch ~/ 1000;
    return '$mentorDocId${MeetingConstants.meetingIdSeparator}$timestamp';
  }
  
  /// Convert Timestamp to ISO string
  static String timestampToIsoString(dynamic timestamp) {
    if (timestamp is Timestamp) {
      return timestamp.toDate().toIso8601String();
    } else if (timestamp is String) {
      return timestamp;
    }
    return '';
  }
  
  /// Convert string time to DateTime for a specific date
  /// Example: "14:00" on "2024-01-15" -> DateTime
  static DateTime? parseTimeSlot(String date, String time) {
    try {
      return DateTime.parse('${date}T$time:00');
    } catch (e) {
      return null;
    }
  }
  
  /// Check if a document should be skipped (metadata documents)
  static bool shouldSkipDocument(String docId) {
    return docId == MeetingConstants.metadataDocName || 
           docId.startsWith(MeetingConstants.metadataDocPrefix);
  }
  
  /// Get week number for a date (ISO week number)
  static int getWeekNumber(DateTime date) {
    // Calculate the Thursday of the week (ISO 8601)
    final thursday = date.add(Duration(days: 4 - date.weekday));
    // Find the first Thursday of the year
    final firstThursday = DateTime(thursday.year, 1, 1);
    final daysToFirstThursday = firstThursday.weekday == DateTime.thursday 
        ? 0 
        : (8 - firstThursday.weekday + DateTime.thursday) % 7;
    final firstThursdayOfYear = firstThursday.add(Duration(days: daysToFirstThursday));
    
    // Calculate week number
    final weekNumber = ((thursday.difference(firstThursdayOfYear).inDays) / 7).floor() + 1;
    return weekNumber;
  }
  
  /// Format date for display
  static String formatDate(DateTime date) {
    return '${date.year}-${date.month.toString().padLeft(2, '0')}-${date.day.toString().padLeft(2, '0')}';
  }
  
  /// Get month-year string for grouping
  static String getMonthYear(DateTime date) {
    return '${date.year}-${date.month.toString().padLeft(2, '0')}';
  }
  
  /// Create a query-friendly compound field
  static String createCompoundField(String field1, String field2) {
    return '${field1}_${field2}';
  }
  
  /// Add one hour to a time string
  static String addHour(String time) {
    final parts = time.split(':').map(int.parse).toList();
    if (parts.length >= 2) {
      final hours = (parts[0] + 1) % 24;
      final minutes = parts[1];
      return '${hours.toString().padLeft(2, '0')}:${minutes.toString().padLeft(2, '0')}';
    }
    return time;
  }
}

File: functions/src/meetings/utils/meeting-helpers.ts
import * as admin from 'firebase-admin';
import { getUniversityCollection } from '../../utils/database';

/**
 * Helper functions for meeting operations
 */

/**
 * Get user document by Firebase UID
 * Returns the document ID and data
 */
export async function getUserDocByUid(
  universityPath: string, 
  uid: string
): Promise<{ id: string; data: FirebaseFirestore.DocumentData } | null> {
  const usersCollection = getUniversityCollection(universityPath, 'users');
  
  // First try direct document ID lookup (in case uid is actually a doc ID)
  try {
    const docById = await usersCollection.doc(uid).get();
    if (docById.exists) {
      return { id: docById.id, data: docById.data()! };
    }
  } catch (e) {
    // Not a valid document ID, continue to firebase_uid lookup
  }
  
  // Query by firebase_uid
  const query = await usersCollection
    .where('firebase_uid', '==', uid)
    .limit(1)
    .get();
  
  if (!query.empty) {
    const doc = query.docs[0];
    return { id: doc.id, data: doc.data() };
  }
  
  return null;
}

/**
 * Generate human-readable meeting ID
 * Format: {mentorDocId}__{menteeDocId}__{timestamp}
 */
export function generateMeetingId(
  mentorDocId: string, 
  menteeDocId: string, 
  startTime: string | Date
): string {
  const timestamp = startTime instanceof Date 
    ? Math.floor(startTime.getTime() / 1000)
    : Math.floor(new Date(startTime).getTime() / 1000);
  
  return `${mentorDocId}__${menteeDocId}__${timestamp}`;
}

/**
 * Generate availability slot ID
 * Format: {mentorDocId}__{timestamp}
 */
export function generateAvailabilityId(
  mentorDocId: string,
  date: string,
  slotStart: string
): string {
  const timestamp = Math.floor(new Date(`${date}T${slotStart}`).getTime() / 1000);
  return `${mentorDocId}__${timestamp}`;
}

/**
 * Parse old availability ID format to get document ID and slot index
 * Format: "docId_slot_0"
 */
export function parseOldAvailabilityId(availabilityId: string): {
  docId: string;
  slotIndex: number;
} | null {
  const parts = availabilityId.split('_slot_');
  if (parts.length !== 2) return null;
  
  const docId = parts[0];
  const slotIndex = parseInt(parts[1]);
  
  if (isNaN(slotIndex)) return null;
  
  return { docId, slotIndex };
}

/**
 * Get week number for a date (ISO week number)
 */
export function getWeekNumber(date: Date): number {
  // Calculate the Thursday of the week (ISO 8601)
  const thursday = new Date(date);
  thursday.setDate(date.getDate() + (4 - date.getDay()));
  
  // Find the first Thursday of the year
  const firstThursday = new Date(thursday.getFullYear(), 0, 1);
  const daysToFirstThursday = firstThursday.getDay() === 4 
    ? 0 
    : (8 - firstThursday.getDay() + 4) % 7;
  firstThursday.setDate(firstThursday.getDate() + daysToFirstThursday);
  
  // Calculate week number
  const weekNumber = Math.floor((thursday.getTime() - firstThursday.getTime()) / (7 * 24 * 60 * 60 * 1000)) + 1;
  return weekNumber;
}

/**
 * Format date as YYYY-MM-DD
 */
export function formatDate(date: Date): string {
  const year = date.getFullYear();
  const month = (date.getMonth() + 1).toString().padStart(2, '0');
  const day = date.getDate().toString().padStart(2, '0');
  return `${year}-${month}-${day}`;
}

/**
 * Get month-year string for grouping
 */
export function getMonthYear(date: Date): string {
  const year = date.getFullYear();
  const month = (date.getMonth() + 1).toString().padStart(2, '0');
  return `${year}-${month}`;
}

/**
 * Add hours to a time string
 */
export function addHours(time: string, hours: number = 1): string {
  const [h, m] = time.split(':').map(Number);
  const newHours = (h + hours) % 24;
  return `${newHours.toString().padStart(2, '0')}:${m.toString().padStart(2, '0')}`;
}